### PR TITLE
Add a triggered with put-completion IOC example + exponential decay IOC example

### PR DESCRIPTION
**Pull request by tacaswell**

### CODE REVIEW

**Comment by klauer** [comment]
I like it. Think we could generalize this and make it a usable primitive? `pvproperty` new option or subclass, perhaps... ?

**Comment by tacaswell** [comment]
@klauer I assume you are talking about the no-re-entry thing?

**Comment by klauer** [comment]
I meant more of a general "acquire"-style signal that does pretty much everything you've done here.

File: caproto/asyncio/server.py [hunk 1]
```diff
@@ -303,6 +303,7 @@ def run(pvdb, *, interfaces=None, log_pv_names=False):
         start_server(pvdb, interfaces=interfaces, log_pv_names=log_pv_names))
     try:
         loop.run_until_complete(task)
-    except KeyboardInterrupt:
+    finally:
+        print('in finally')
```
Comment (danielballan) [inline caproto/asyncio/server.py]: Stray print.

**Comment by klauer** [comment]
It'd be nice to pull out this example + split the fixes/server changes.

**Comment by tacaswell** [comment]
I squashed this and re-ordered it.  Can pop the last 3 commits off of this into a different PR if folks want.

**Comment by klauer** [comment]
OK, since @danielballan put the server fixes in a separate PR, would you mind popping those commits off @tacaswell?

File: caproto/ioc_examples/mini_beamline.py [hunk 2]
```diff
@@ -21,14 +39,34 @@ class _JitterDetector(PVGroup):
...
     exp = pvproperty(value=1, dtype=float)
 
     @exp.putter
     async def exp(self, instance, value):
         value = np.clip(value, a_min=0, a_max=None)
         return value
 
+    @mtr.startup
+    async def mtr(self, instance, async_lib):
+        print('here')
```
Comment (klauer) [inline caproto/ioc_examples/mini_beamline.py]:
```suggestion
```

File: caproto/ioc_examples/trigger_with_pc.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    @wait.getter
+    async def wait(self, instance):
+        sleep_time = random.randint(0, 15)
+        await self.wait_time.write(sleep_time)
+        await instance.async_lib.library.sleep(sleep_time)
+        return sleep_time
+
+    fatal = pvproperty(value=0)
+
+    @fatal.getter
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply caproto/ioc_examples/trigger_with_pc.py]: ?

[the remark this answers is not in the document]
Comment (danielballan) [reply caproto/ioc_examples/trigger_with_pc.py]: What is the use case for PV that kills the server when you read it? I can imagine for test/teaching purposes wanting a way for a client to knock a sever over, but would it be more realistic and just as useful to do it on write?

[the remark this answers is not in the document]
Comment (tacaswell) [reply caproto/ioc_examples/trigger_with_pc.py]: I have no memory of why I did it this way.

File: caproto/ioc_examples/decay.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#!/USSR/bin/env python3
```
[the remark this answers is not in the document]
Comment (danielballan) [reply caproto/ioc_examples/decay.py]: Amazing. This wasn't covered on the "export control" training that I just renewed.

**Comment by danielballan** [comment]
It looks like this is about ready. I have push-forced a rebase (to pick up precommit tooling) and added a commit to fix a sphinx error.

File: caproto/ioc_examples/trigger_with_pc.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielballan) [reply caproto/ioc_examples/trigger_with_pc.py]: Fixed in  bd3b176

**Comment by danielballan** [comment]
Also it is useful to know that `async_lib.libary.Event` has varied behavior (curio and trio/asyncio disagree about which methods are awaitable) whereas `async_lib.Event` is wrapped (by us) in a standard API.

**Comment by klauer** [comment]
There's still some CI env issues:

```
++ EPICS_CA_ADDR_LIST=127.255.255.255
++ EPICS_CA_AUTO_ADDR_LIST=NO
Set EPICS_CA_ADDR_LIST to 127.255.255.255
```

AUTO_ADDR_LIST should be yes

**Comment by danielballan** [comment]
@ke-zhang-rd These are env issues in the Azure builds?

**Comment by tacaswell** [comment]
maybe rebase on master?

**Comment by danielballan** [comment]
I just rebased and force-pushed this branch this morning. We're still good:

```
14:04 $ git pull tacaswell triggered_ioc 
From github.com:tacaswell/caproto
 * branch              triggered_ioc -> FETCH_HEAD
Already up to date.
(py382) ✔ ~/Repos/bnl/caproto [triggered_ioc|…1⚑ 26] 
14:04 $ git rebase master
Current branch triggered_ioc is up to date.
```

**Comment by klauer** [comment]
I thought it was a rebase issue, but now I think it's just another indicator our CI needs an overhaul...

**Comment by danielballan** [comment]
@MikeHart85 has started surveying what's out there (VMs, ansible playbooks, Dockerfiles, epics-of-travis, etc.) and that might eventually lead to a collaborative CI overhaul, among other things.

**Pull request merged**